gcp-client/src/main/java/com/google/graphite/platforms/plugin/client/ComputeClient.java

```diff
@@ -0,0 +1,541 @@
+
+  public List<InstanceTemplate> getTemplates(String projectId) throws IOException {
+    List<InstanceTemplate> instanceTemplates =
+        compute.instanceTemplates().list(projectId).execute().getItems();
+    if (instanceTemplates == null) {
+      instanceTemplates = Collections.emptyList();
+    }
+
+    // Sort by name
+    instanceTemplates.sort(Comparator.comparing(InstanceTemplate::getName));
+
+    return instanceTemplates;
+  }
+
+  /**
+   * Creates persistent disk snapshot for Compute Engine instance. This method blocks until the
+   * operation completes.
+   *
+   * @param projectId Google cloud project id (e.g. my-project).
+   * @param zone Instance's zone.
+   * @param instanceId Name of the instance whose disks to take a snapshot of.
+   * @throws IOException If an error occured in snapshot creation.
+   * @throws InterruptedException If snapshot creation is interrupted.
+   */
+  public void createSnapshot(String projectId, String zone, String instanceId)
+      throws IOException, InterruptedException {
+    try {
+      zone = nameFromSelfLink(zone);
+      Instance instance = compute.instances().get(projectId, zone, instanceId).execute();
+
+      // TODO: JENKINS-56113 parallelize snapshot creation
+      for (AttachedDisk disk : instance.getDisks()) {
+        String diskId = nameFromSelfLink(disk.getSource());
+        createSnapshotForDisk(projectId, zone, diskId);
+      }
+    } catch (InterruptedException ie) {
+      // catching InterruptedException here because calling function also can throw
+      // InterruptedException from trying to terminate node
+      LOGGER.log(Level.WARNING, "Error in creating snapshot.", ie);
+      throw ie;
+    } catch (IOException ioe) {
+      LOGGER.log(Level.WARNING, "Interruption in creating snapshot.", ioe);
+      throw ioe;
+    }
+  }
+
+  /**
+   * Given a disk's name, create a snapshot for said disk.
+   *
+   * @param projectId Google cloud project id.
+   * @param zone Zone of disk.
+   * @param diskId Name of disk to create a snapshot for.
+   * @throws IOException If an error occured in snapshot creation.
+   * @throws InterruptedException If snapshot creation is interrupted.
+   */
+  public void createSnapshotForDisk(String projectId, String zone, String diskId)
+      throws IOException, InterruptedException {
+    Snapshot snapshot = new Snapshot();
+    snapshot.setName(diskId);
+
+    Operation op = compute.disks().createSnapshot(projectId, zone, diskId, snapshot).execute();
+    // poll for result
+    waitForOperationCompletion(
+        projectId, op.getName(), op.getZone(), SNAPSHOT_TIMEOUT_MILLISECONDS);
+  }
+
+  /**
+   * Deletes persistent disk snapshot. Does not block.
+   *
+   * @param projectId Google cloud project id.
+   * @param snapshotName Name of the snapshot to be deleted.
+   * @throws IOException If an error occurred in deleting the snapshot.
+   */
+  public void deleteSnapshot(String projectId, String snapshotName) throws IOException {
```

Comment:
Suggest Async method name suffix in this case. Also suggest returning the operation to give downstream callers the option to verify/wait.

Reply:
Also suggest returning an operation anywhere else there's just a void return type on an API call like this. In general, the more flexibility we can bake into this library now, the less re-factoring we'll have to do later.

Reply:
Async suffix: Done.

void return types: I've done this where possible. This becomes a lot messier when parallelizing snapshot creation, so I've left that method with the void return type for now.